lib/stm32/g0/flash.c

```diff
@@ -30,6 +30,145 @@
 
 #include <libopencm3/stm32/flash.h>
 
+/** @brief Wait until Last Flash Operation has Ended */
+void flash_wait_for_last_operation(void)
+{
+	while ((FLASH_SR & FLASH_SR_BSY) == FLASH_SR_BSY);
+}
+
+/** @brief Program a 64bits word to FLASH
+ *
+ * Program a 64bit word to FLASH memory.
+ * Flash programing error must be checked and cleared if needed
+ * before starting programming.
+ *
+ * @param[in] address Address in FLASH
+ * @param[in] data Double word to write
+ */
+void flash_program_double_word(uint32_t address, uint64_t data)
+{
+	flash_wait_for_last_operation();
+
+	FLASH_CR |= FLASH_CR_PG;
+
+	MMIO32(address) = (uint32_t)data;
+	MMIO32(address+4) = (uint32_t)(data >> 32);
+
+	flash_wait_for_last_operation();
+
+	FLASH_CR &= ~FLASH_CR_PG;
+}
+
+/** @brief Program a Data Block to FLASH
+ *
+ * This programs an arbitrary length data block to FLASH memory.
+ * The program error flag should be checked separately for the event that memory
+ * was not properly erased.
+ *
+ * @param[in] address Starting address in Flash.
+ * @param[in] data Pointer to start of data block.
+ * @param[in] len Length of data block in bytes (multiple of 8).
+ **/
+void flash_program(uint32_t address, uint8_t *data, uint32_t len)
+{
+	for (uint32_t i = 0; i < len; i += 8) {
+		flash_program_double_word(address+i, *(uint64_t*)(data + i));
+	}
+}
+
+/** @brief Erase a page of FLASH
+ * @param[in] page
+ */
+void flash_erase_page(uint32_t page)
+{
+	flash_wait_for_last_operation();
+
+	uint32_t reg = FLASH_CR;
+	reg &= (FLASH_CR_PNB_SHIFT << FLASH_CR_PNB_SHIFT);
```

Comment:
that doesn't look correct.  you want an ~ and a mask right?

Reply:
ow ..